### PR TITLE
Ft user logout 153081281

### PR DESCRIPTION
**Pull request by mirr254**
**What does this PR do?**
Implements logout functionality

**Description of tasks completed**
- Creating a BlacklistToken model
- adds test for logout
- implement logout endpoint by inserting the token to `blacklisted_tokens` table
- Put access control to secured endpoints. Blacklisted tokens are considered expired

**How should this PR be manually tested**
- clone the repo locally and install dependencies from `requirements.txt`
- `checkout` to `ft-user-logout-153081281` branch and run the app
- use postman or curl to test the endpoint `/api/v1/auth/login` 
- logout same user using `/api/v1/auth/logout`

**Related pivotal stories**
[153081281](https://www.pivotaltracker.com/story/show/153081281

**Screenshots**
logout

restrictaccess

### CODE REVIEW

File: tests/test_user_auth.py [hunk 1]
```diff
@@ -65,22 +90,39 @@ def open_with_auth(self, url, method, username, password):
...
-        # self.assertEqual(res.status_code, 201)
+
         self.test_auth_register()
-        res = self.open_with_auth('/api/v1/auth/login', 'GET', 'test', 'hardpass')
-        
+        res = self.auth_login()
+        self.assertEqual(res.status_code, 200)
+
+    #test for logout endpoint
+    def test_auth_logout(self):
```
Comment (meshnesh) [inline tests/test_user_auth.py:99]: Good implementation of the test, although make the test atomic

File: tests/test_user_auth.py [hunk 2]
```diff
@@ -50,10 +50,35 @@ def setUp(self):
...
+    def auth_login(self):
+        return self.open_with_auth('/api/v1/auth/login', 'GET', 'test', 'hardpass')
+
+    #register a user
+    def register_users(self):
+        user_details = json.dumps(self.user)
+        return self.client().post('api/v1/auth/register', data=json.dumps(self.user), content_type='application/json')
+
+     #get the token from logged in user
+    def get_verfication_token(self):
```
Comment (meshnesh) [inline tests/test_user_auth.py:66]: good implementation of the token handler

File: app/auth_blueprint/models.py [hunk 3]
```diff
@@ -26,4 +26,24 @@ def __repr__(self):
         return "<User: {}>".format(self.username) #object instance of the model whenever it is queried
 
 
+class TokenBlackList(db.Model):
```
Comment (meshnesh) [inline app/auth_blueprint/models.py:29]: Nice implementation of blacklisting tokens when user logout, although why do you need a column of `date_created` when black listing?

File: app/auth_blueprint/views.py [hunk 4]
```diff
@@ -102,5 +101,8 @@ def reset_password(email):
 @auth.route('/api/v1/auth/logout')
 @token_required
 def logout(logged_in_user):
```
Comment (meshnesh) [inline app/auth_blueprint/views.py:103]: good implementation of the logout

File: app/events_blueprint/views.py [hunk 5]
```diff
@@ -2,50 +2,54 @@
...
 
 """ HANDLE EVENTS ACTIVITIES """
 #create a new event
 @events.route('/api/v1/events', methods=['POST'])
 @token_required
-def add_event(logged_in_user):
-    #check if user is logged in
+def add_event(logged_in_user):    
+     #check if token is blacklisted
+    token = request.headers['x-access-token']
```
Comment (meshnesh) [inline app/events_blueprint/views.py]: if you have a decorator that handles the requirements of the token, why check for it again here?

**Review by meshnesh — COMMENTED** [review]
Since you have a decorator for the token, It would be a best practice to use it to handle all the token requirements.
Try implementing DRY principle, which is unfollowed in most case for the token.

File: app/auth_blueprint/models.py (hunk 3 above)
Comment (mirr254) [reply app/auth_blueprint/models.py:29]: Not necessary but keeping track of the logout time

File: app/events_blueprint/views.py (hunk 5 above)
Comment (mirr254) [reply app/events_blueprint/views.py]: The decorator checks for expired tokens and invalid ones. Doesn't check for blacklisted tokens. Maybe I incorporate that in the decorator and remove the blacklist check from all endpoints used. Thanks, I hadn't thought of that.

**Pull request merged**